File: sdk/typescript/src/wire.ts
// Binary wire protocol implementation for AxiomVM appliance communication.

export const PROTOCOL_MAGIC = 0x4158494f;
export const PROTOCOL_VERSION = 0x0001;
export const HEADER_SIZE = 20;

export const EXEC_FLAG_STRIP_ANSI = 1 << 0;
export const EXEC_FLAG_CAPTURE_DIFF = 1 << 1;

export enum Opcode {
  PING = 0x0001,
  PONG = 0x0002,
  EXEC = 0x0010,
  EXEC_RESPONSE = 0x0011,
  AST_SYMBOLS = 0x0020,
  AST_SYMBOLS_RESPONSE = 0x0021,
  AST_SLICE = 0x0022,
  AST_SLICE_RESPONSE = 0x0023,
  AST_PATCH = 0x0024,
  AST_PATCH_RESPONSE = 0x0025,
  CDP_ACTION = 0x0030,
  CDP_ACTION_RESPONSE = 0x0031,
}

export enum CdpActionType {
  NAVIGATE = 0,
  CLICK = 1,
  TYPE_TEXT = 2,
  SCROLL = 3,
  GET_TREE = 4,
}

export interface Header {
  magic: number;
  version: number;
  opcode: Opcode;
  requestId: bigint;
  payloadLen: number;
}

export interface ExecRequest {
  command: string;
  cwd?: string;
  timeoutMs?: number;
  stripAnsi?: boolean;
  captureDiff?: boolean;
}

export interface ExecResponse {
  exitCode: number;
  flags: number;
  cpuCycles: bigint;
  output: string;
  diff: string;
}

export interface AstSymbolsResponse {
  status: number;
  symbolCount: number;
  symbolsText: string;
}

export interface AstSliceResponse {
  status: number;
  startLine: number;
  endLine: number;
  content: string;
}

export interface AstPatchResponse {
  status: number;
  oldStartLine: number;
  oldEndLine: number;
  newEndLine: number;
}

export interface CdpActionRequest {
  action: CdpActionType;
  flags?: number;
  targetId?: number;
  scrollDelta?: number;
  payload1?: string;
  payload2?: string;
}

export interface CdpActionResponse {
  status: number;
  nodeCount: number;
  treeText: string;
}

// Encodes a standard 20-byte wire protocol header.
export function encodeHeader(
  opcode: Opcode,
  requestId: bigint,
  payloadLen: number
): Buffer {
  const buf = Buffer.alloc(HEADER_SIZE);
  buf.writeUInt32LE(PROTOCOL_MAGIC, 0);
  buf.writeUInt16LE(PROTOCOL_VERSION, 4);
  buf.writeUInt16LE(opcode, 6);
  buf.writeBigUInt64LE(requestId, 8);
  buf.writeUInt32LE(payloadLen, 16);
  return buf;
}

// Decodes a 20-byte wire protocol header from buffer.
export function decodeHeader(buf: Buffer): Header {
  if (buf.length < HEADER_SIZE) {
    throw new Error(`Buffer underflow: expected ${HEADER_SIZE} bytes`);
  }
  const magic = buf.readUInt32LE(0);
  if (magic !== PROTOCOL_MAGIC) {
    throw new Error(`Invalid protocol magic: 0x${magic.toString(16)}`);
  }
  const version = buf.readUInt16LE(4);
  if (version !== PROTOCOL_VERSION) {
    throw new Error(`Unsupported protocol version: ${version}`);
  }
  const opcode = buf.readUInt16LE(6) as Opcode;
  const requestId = buf.readBigUInt64LE(8);
  const payloadLen = buf.readUInt32LE(16);
  return { magic, version, opcode, requestId, payloadLen };
}

// Encodes a command execution request payload.
export function encodeExecRequest(req: ExecRequest): Buffer {
  const commandBuf = Buffer.from(req.command, "utf-8");
  const cwdBuf = Buffer.from(req.cwd ?? "/tmp", "utf-8");
  let flags = 0;
  if (req.stripAnsi ?? true) flags |= EXEC_FLAG_STRIP_ANSI;
  if (req.captureDiff ?? true) flags |= EXEC_FLAG_CAPTURE_DIFF;
  const timeoutMs = req.timeoutMs ?? 30000;

  const header = Buffer.alloc(16);
  header.writeUInt32LE(flags, 0);
  header.writeUInt32LE(timeoutMs, 4);
  header.writeUInt32LE(commandBuf.length, 8);
  header.writeUInt32LE(cwdBuf.length, 12);
  return Buffer.concat([header, commandBuf, cwdBuf]);
}

// Decodes an execution response payload into an ExecResponse.
export function decodeExecResponse(buf: Buffer): ExecResponse {
  if (buf.length < 24) {
    throw new Error("Payload underflow decoding ExecResponse");
  }
  const exitCode = buf.readInt32LE(0);
  const flags = buf.readUInt32LE(4);
  const cpuCycles = buf.readBigUInt64LE(8);
  const outputLen = buf.readUInt32LE(16);
  const diffLen = buf.readUInt32LE(20);

  if (buf.length < 24 + outputLen + diffLen) {
    throw new Error("Payload underflow for output and diff segments");
  }
  const output = buf.subarray(24, 24 + outputLen).toString("utf-8");
  const diff = buf
    .subarray(24 + outputLen, 24 + outputLen + diffLen)
    .toString("utf-8");
  return { exitCode, flags, cpuCycles, output, diff };
}

// Encodes an AST symbol outline request payload.
export function encodeAstSymbolsRequest(filePath: string): Buffer {
  const pathBuf = Buffer.from(filePath, "utf-8");
  const hdr = Buffer.alloc(4);
  hdr.writeUInt32LE(pathBuf.length, 0);
  return Buffer.concat([hdr, pathBuf]);
}

// Decodes an AST symbol outline response payload.
export function decodeAstSymbolsResponse(buf: Buffer): AstSymbolsResponse {
  if (buf.length < 12) {
    throw new Error("Payload underflow decoding AstSymbolsResponse");
  }
  const status = buf.readInt32LE(0);
  const symbolCount = buf.readUInt32LE(4);
  const payloadLen = buf.readUInt32LE(8);
  if (buf.length < 12 + payloadLen) {
    throw new Error("Payload underflow for symbols text");
  }
  const symbolsText = buf.subarray(12, 12 + payloadLen).toString("utf-8");
  return { status, symbolCount, symbolsText };
}

// Encodes an AST symbol slice extraction request payload.
export function encodeAstSliceRequest(
  filePath: string,
  symbol: string
): Buffer {
  const pathBuf = Buffer.from(filePath, "utf-8");
  const symBuf = Buffer.from(symbol, "utf-8");
  const hdr = Buffer.alloc(4);
  hdr.writeUInt16LE(pathBuf.length, 0);
  hdr.writeUInt16LE(symBuf.length, 2);
  return Buffer.concat([hdr, pathBuf, symBuf]);
}

// Decodes an AST symbol slice response payload.
export function decodeAstSliceResponse(buf: Buffer): AstSliceResponse {
  if (buf.length < 16) {
    throw new Error("Payload underflow decoding AstSliceResponse");
  }
  const status = buf.readInt32LE(0);
  const startLine = buf.readUInt32LE(4);
  const endLine = buf.readUInt32LE(8);
  const contentLen = buf.readUInt32LE(12);
  if (buf.length < 16 + contentLen) {
    throw new Error("Payload underflow for slice content");
  }
  const content = buf.subarray(16, 16 + contentLen).toString("utf-8");
  return { status, startLine, endLine, content };
}

// Encodes an AST symbol replacement patch request payload.
export function encodeAstPatchRequest(
  filePath: string,
  symbol: string,
  replacement: string
): Buffer {
  const pathBuf = Buffer.from(filePath, "utf-8");
  const symBuf = Buffer.from(symbol, "utf-8");
  const repBuf = Buffer.from(replacement, "utf-8");
  const hdr = Buffer.alloc(8);
  hdr.writeUInt16LE(pathBuf.length, 0);
  hdr.writeUInt16LE(symBuf.length, 2);
  hdr.writeUInt32LE(repBuf.length, 4);
  return Buffer.concat([hdr, pathBuf, symBuf, repBuf]);
}

// Decodes an AST symbol replacement patch response payload.
export function decodeAstPatchResponse(buf: Buffer): AstPatchResponse {
  if (buf.length < 16) {
    throw new Error("Payload underflow decoding AstPatchResponse");
  }
  const status = buf.readInt32LE(0);
  const oldStartLine = buf.readUInt32LE(4);
  const oldEndLine = buf.readUInt32LE(8);
  const newEndLine = buf.readUInt32LE(12);
  return { status, oldStartLine, oldEndLine, newEndLine };
}

// Encodes a browser automation CDP action request payload.
export function encodeCdpRequest(req: CdpActionRequest): Buffer {
  const p1Buf = Buffer.from(req.payload1 ?? "", "utf-8");
  const p2Buf = Buffer.from(req.payload2 ?? "", "utf-8");
  const hdr = Buffer.alloc(20);
  hdr.writeUInt16LE(req.action, 0);
  hdr.writeUInt16LE(req.flags ?? 0, 2);
  hdr.writeUInt32LE(req.targetId ?? 0, 4);
  hdr.writeInt32LE(req.scrollDelta ?? 0, 8);
  hdr.writeUInt32LE(p1Buf.length, 12);
  hdr.writeUInt32LE(p2Buf.length, 16);
  return Buffer.concat([hdr, p1Buf, p2Buf]);
}

// Decodes a browser automation CDP action response payload.
export function decodeCdpResponse(buf: Buffer): CdpActionResponse {
  if (buf.length < 12) {
    throw new Error("Payload underflow decoding CdpActionResponse");
  }
  const status = buf.readInt32LE(0);
  const nodeCount = buf.readUInt32LE(4);
  const payloadLen = buf.readUInt32LE(8);
  if (buf.length < 12 + payloadLen) {
    throw new Error("Payload underflow for tree text");
  }
  const treeText = buf.subarray(12, 12 + payloadLen).toString("utf-8");
  return { status, nodeCount, treeText };
}

File: host/axiom/wire.py
"""TDS-Wire protocol codec for host orchestration."""

import struct
from dataclasses import dataclass
from enum import IntEnum

WIRE_MAGIC = 0xAA55
HEADER_SIZE = 12
HEADER_FORMAT = "<HHII"


class Opcode(IntEnum):
    """Operation codes supported by TDS-Wire protocol."""

    PING = 0x0001
    PONG = 0x0002
    EXEC = 0x0010
    EXEC_OUTPUT = 0x0011
    AST_SYMBOLS = 0x0020
    AST_SLICE = 0x0021
    AST_PATCH = 0x0022
    AST_SYMBOLS_RESPONSE = 0x0023
    AST_SLICE_RESPONSE = 0x0024
    AST_PATCH_RESPONSE = 0x0025
    CDP_ACTION = 0x0030
    CDP_ACTION_RESPONSE = 0x0031
    SHUTDOWN = 0x00FF


@dataclass(frozen=True, slots=True)
class FrameHeader:
    """Wire protocol header structure matching C++ memory topology."""

    magic: int
    opcode: int
    request_id: int
    payload_len: int


def encode_frame(opcode: int, request_id: int, payload: bytes = b"") -> bytes:
    """Encodes a TDS-Wire frame with header and optional payload bytes."""
    header = struct.pack(HEADER_FORMAT, WIRE_MAGIC, opcode, request_id, len(payload))
    return header + payload


def decode_header(buffer: bytes) -> FrameHeader:
    """Decodes a frame header from raw buffer bytes."""
    if len(buffer) < HEADER_SIZE:
        raise ValueError("Buffer underflow decoding frame header")
    magic, opcode, request_id, payload_len = struct.unpack(
        HEADER_FORMAT, buffer[:HEADER_SIZE]
    )
    if magic != WIRE_MAGIC:
        raise ValueError(f"Invalid magic 0x{magic:04X}, expected 0x{WIRE_MAGIC:04X}")
    return FrameHeader(
        magic=magic, opcode=opcode, request_id=request_id, payload_len=payload_len
    )


EXEC_FLAG_STRIP_ANSI = 1 << 0
EXEC_FLAG_CAPTURE_DIFF = 1 << 1

EXEC_RESP_TRUNCATED = 1 << 0
EXEC_RESP_TIMED_OUT = 1 << 1

EXEC_REQ_FORMAT = "<IIHH"
EXEC_REQ_HEADER_SIZE = 12

EXEC_RESP_FORMAT = "<iIQII"
EXEC_RESP_HEADER_SIZE = 24


@dataclass(frozen=True, slots=True)
class ExecRequest:
    """Request parameters for command execution inside appliance."""

    command: str
    cwd: str = ""
    timeout_ms: int = 5000
    flags: int = EXEC_FLAG_STRIP_ANSI | EXEC_FLAG_CAPTURE_DIFF


@dataclass(frozen=True, slots=True)
class ExecResponse:
    """Execution response returned by appliance daemon."""

    exit_code: int
    flags: int
    cpu_cycles: int
    output: str
    diff: str


def encode_exec_request(req: ExecRequest) -> bytes:
    """Serializes an ExecRequest into binary payload bytes."""
    cmd_bytes = req.command.encode("utf-8")
    cwd_bytes = req.cwd.encode("utf-8")
    hdr = struct.pack(
        EXEC_REQ_FORMAT,
        req.flags,
        req.timeout_ms,
        len(cmd_bytes),
        len(cwd_bytes),
    )
    return hdr + cmd_bytes + cwd_bytes


def decode_exec_response(payload: bytes) -> ExecResponse:
    """Parses binary response payload into an ExecResponse dataclass."""
    if len(payload) < EXEC_RESP_HEADER_SIZE:
        raise ValueError("Payload underflow decoding ExecResponse")
    exit_code, flags, cpu_cycles, output_len, diff_len = struct.unpack(
        EXEC_RESP_FORMAT, payload[:EXEC_RESP_HEADER_SIZE]
    )
    data = payload[EXEC_RESP_HEADER_SIZE:]
    if len(data) < output_len + diff_len:
        raise ValueError("Payload underflow for output and diff segments")
    output_bytes = data[:output_len]
    diff_bytes = data[output_len : output_len + diff_len]
    return ExecResponse(
        exit_code=exit_code,
        flags=flags,
        cpu_cycles=cpu_cycles,
        output=output_bytes.decode("utf-8", errors="replace"),
        diff=diff_bytes.decode("utf-8", errors="replace"),
    )


AST_SYMBOLS_REQ_FORMAT = "<I"
AST_SYMBOLS_REQ_HEADER_SIZE = 4
AST_SYMBOLS_RESP_FORMAT = "<iII"
AST_SYMBOLS_RESP_HEADER_SIZE = 12

AST_SLICE_REQ_FORMAT = "<HH"
AST_SLICE_REQ_HEADER_SIZE = 4
AST_SLICE_RESP_FORMAT = "<iIII"
AST_SLICE_RESP_HEADER_SIZE = 16

AST_PATCH_REQ_FORMAT = "<HHI"
AST_PATCH_REQ_HEADER_SIZE = 8
AST_PATCH_RESP_FORMAT = "<iIII"
AST_PATCH_RESP_HEADER_SIZE = 16


@dataclass(frozen=True, slots=True)
class AstSymbolsRequest:
    """Request parameters for querying symbol declarations in a file."""

    path: str


@dataclass(frozen=True, slots=True)
class AstSymbolsResponse:
    """Structured symbol outline returned from AST analysis."""

    status: int
    symbol_count: int
    symbols_text: str


def encode_ast_symbols_request(req: AstSymbolsRequest) -> bytes:
    """Serializes an AstSymbolsRequest into binary payload bytes."""
    path_bytes = req.path.encode("utf-8")
    hdr = struct.pack(AST_SYMBOLS_REQ_FORMAT, len(path_bytes))
    return hdr + path_bytes


def decode_ast_symbols_response(payload: bytes) -> AstSymbolsResponse:
    """Parses binary response payload into an AstSymbolsResponse dataclass."""
    if len(payload) < AST_SYMBOLS_RESP_HEADER_SIZE:
        raise ValueError("Payload underflow decoding AstSymbolsResponse")
    status, symbol_count, payload_len = struct.unpack(
        AST_SYMBOLS_RESP_FORMAT, payload[:AST_SYMBOLS_RESP_HEADER_SIZE]
    )
    data = payload[AST_SYMBOLS_RESP_HEADER_SIZE:]
    if len(data) < payload_len:
        raise ValueError("Payload underflow for symbols text")
    symbols_text = data[:payload_len].decode("utf-8", errors="replace")
    return AstSymbolsResponse(
        status=status, symbol_count=symbol_count, symbols_text=symbols_text
    )


@dataclass(frozen=True, slots=True)
class AstSliceRequest:
    """Request parameters for extracting the source slice of a symbol."""

    path: str
    symbol: str


@dataclass(frozen=True, slots=True)
class AstSliceResponse:
    """Symbol slice content and line boundaries returned from AST analysis."""

    status: int
    start_line: int
    end_line: int
    content: str


def encode_ast_slice_request(req: AstSliceRequest) -> bytes:
    """Serializes an AstSliceRequest into binary payload bytes."""
    path_bytes = req.path.encode("utf-8")
    sym_bytes = req.symbol.encode("utf-8")
    hdr = struct.pack(AST_SLICE_REQ_FORMAT, len(path_bytes), len(sym_bytes))
    return hdr + path_bytes + sym_bytes


def decode_ast_slice_response(payload: bytes) -> AstSliceResponse:
    """Parses binary response payload into an AstSliceResponse dataclass."""
    if len(payload) < AST_SLICE_RESP_HEADER_SIZE:
        raise ValueError("Payload underflow decoding AstSliceResponse")
    status, start_line, end_line, content_len = struct.unpack(
        AST_SLICE_RESP_FORMAT, payload[:AST_SLICE_RESP_HEADER_SIZE]
    )
    data = payload[AST_SLICE_RESP_HEADER_SIZE:]
    if len(data) < content_len:
        raise ValueError("Payload underflow for slice content")
    content = data[:content_len].decode("utf-8", errors="replace")
    return AstSliceResponse(
        status=status, start_line=start_line, end_line=end_line, content=content
    )


@dataclass(frozen=True, slots=True)
class AstPatchRequest:
    """Request parameters for patching a symbol with replacement code."""

    path: str
    symbol: str
    replacement: str


@dataclass(frozen=True, slots=True)
class AstPatchResponse:
    """Result of structural symbol replacement operation."""

    status: int
    old_start_line: int
    old_end_line: int
    new_end_line: int


def encode_ast_patch_request(req: AstPatchRequest) -> bytes:
    """Serializes an AstPatchRequest into binary payload bytes."""
    path_bytes = req.path.encode("utf-8")
    sym_bytes = req.symbol.encode("utf-8")
    rep_bytes = req.replacement.encode("utf-8")
    hdr = struct.pack(
        AST_PATCH_REQ_FORMAT, len(path_bytes), len(sym_bytes), len(rep_bytes)
    )
    return hdr + path_bytes + sym_bytes + rep_bytes


def decode_ast_patch_response(payload: bytes) -> AstPatchResponse:
    """Parses binary response payload into an AstPatchResponse dataclass."""
    if len(payload) < AST_PATCH_RESP_HEADER_SIZE:
        raise ValueError("Payload underflow decoding AstPatchResponse")
    status, old_start_line, old_end_line, new_end_line = struct.unpack(
        AST_PATCH_RESP_FORMAT, payload[:AST_PATCH_RESP_HEADER_SIZE]
    )
    return AstPatchResponse(
        status=status,
        old_start_line=old_start_line,
        old_end_line=old_end_line,
        new_end_line=new_end_line,
    )


class CdpActionType(IntEnum):
    """Supported browser action types for CDP interaction."""

    NAVIGATE = 1
    GET_TREE = 2
    CLICK = 3
    TYPE = 4
    SCROLL = 5


CDP_REQ_FORMAT = "<BBHiHH"
CDP_REQ_HEADER_SIZE = 12
CDP_RESP_FORMAT = "<iII"
CDP_RESP_HEADER_SIZE = 12


@dataclass(frozen=True, slots=True)
class CdpActionRequest:
    """Request parameters for browser CDP actions."""

    action: int
    target_id: int = 0
    scroll_delta: int = 0
    payload1: str = ""
    payload2: str = ""
    flags: int = 0


@dataclass(frozen=True, slots=True)
class CdpActionResponse:
    """Result of browser CDP action returned by appliance daemon."""

    status: int
    node_count: int
    tree_text: str


def encode_cdp_request(req: CdpActionRequest) -> bytes:
    """Serializes a CdpActionRequest into binary payload bytes."""
    p1_bytes = req.payload1.encode("utf-8")
    p2_bytes = req.payload2.encode("utf-8")
    hdr = struct.pack(
        CDP_REQ_FORMAT,
        req.action,
        req.flags,
        req.target_id,
        req.scroll_delta,
        len(p1_bytes),
        len(p2_bytes),
    )
    return hdr + p1_bytes + p2_bytes


def decode_cdp_response(payload: bytes) -> CdpActionResponse:
    """Parses binary response payload into a CdpActionResponse dataclass."""
    if len(payload) < CDP_RESP_HEADER_SIZE:
        raise ValueError("Payload underflow decoding CdpActionResponse")
    status, node_count, payload_len = struct.unpack(
        CDP_RESP_FORMAT, payload[:CDP_RESP_HEADER_SIZE]
    )
    data = payload[CDP_RESP_HEADER_SIZE:]
    if len(data) < payload_len:
        raise ValueError("Payload underflow for tree text")
    tree_text = data[:payload_len].decode("utf-8", errors="replace")
    return CdpActionResponse(
        status=status,
        node_count=node_count,
        tree_text=tree_text,
    )
